feat: use github releases to decide on when to build

AtzeDeVries opened the pull request:
This will change the when to build what. 

This will do a diff based on a git tag. When a build has been done on main that builds an oci it will set a tag. Next time on main it will do a diff based on this tag. If the diff actually has changes related to this tag. 

### Why?

We used to do this based on the diff which came with the PR event. This PR event was also used on the main branch. If the job is canceled or removed from queue this creates a problem. Also in github actions, jobs in a queue don't have a guaranteed order. Also rerunning a job based on a PR event would rerun certain steps based on the event data. This is not the behavior you want if the artifact is already published. Also the PR invent was a bit hard to get insight to. 


### What will happen here

When a PR is created the, a diff between the main branch and the PR branch is made and based on the files in the diff decisions are made on which part of CI are required to run.
When merged, mage will look for the most recent tag based on a tag pattern and use this tag as a diff target. When this results in a diff, mage will run certain CI (mage ci for golang with OCI build) and when finished write a new tag. So if you rerun this job it won't rerun because the diff between the tag and the head is none. 
This, in the end will give a way more predictable CI. We can also merge the job queue on the main branch resulting less action minutes spend and not being exposed to the order in which jobs run on the main branch.


### How to use this

If set `inputs.tag-based-diff` is set to `true` this behavior is enabled. This is disabled by default to ensure a smooth transition. It is strongly encouraged to enable this. 

### Updating your image tags in helm

When a release is created we can trigger a job to update the OCI tags in your helm values. This is best done via renovate. See examples below. 


Relevant issues https://github.com/coopnorge/mage/issues/356

AtzeDeVries committed:
feat: use git tags to decide on when to build

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
update permissions

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix typo in command

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
dont depend on github head ref

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix typo in echo

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
set git config

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
also push tag

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
tag with actual user token

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
build auth url for tags

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
use action to push tags

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
also set tags to true

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
revert to native git push

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
try to commit with writeback app

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
bump permissions

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
try to isolate checkout

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fetch right location of json

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
do not sign

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
also use non mage oci prefix tag

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
dont push OCI tag anymore

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
also add mage/oci tag to OCI

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
save it..

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
use allowed tag chars

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
Revert "use allowed tag chars"

This reverts commit 030ef9aecc46fde1e371adfff8e46d2bc939fa11.

AtzeDeVries committed:
Revert "save it.."

This reverts commit a54e6af57d00131fa03091a4053966742de63095.

AtzeDeVries committed:
Revert "also add mage/oci tag to OCI"

This reverts commit d6c40bfdfce20c927a43557b18a216727497b832.

AtzeDeVries committed:
make actuall releases

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
also set predifined tag

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
remoe workdir

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
this would require some disallowed words

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
update target match

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
cleanup comment

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
make using releases and tag based diff conditional

defaults for now to false

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
add docs

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix docs

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

bendiknesbo commented on docs/index.md:
```diff
@@ -64,8 +64,8 @@ other relevant targets for the tech stack.
 import (
-	//mage:import
-	_ "github.com/coopnorge/mage/targets/goapp"
+ //mage:import
```
Incorrect changes to formatting here

AtzeDeVries replied:
not sure really what was going on here, the diff is still here. Might be a hidden char or something

bendiknesbo replied:
Space vs tabs, it looks like

bendiknesbo commented on .github/workflows/reusable-goapp.yaml:
```diff
@@ -25,6 +25,11 @@ on:
+        type: boolean
+        default: false
+        required: false
+        description: If true it will create a github release based on the tag of the OCI
```
I guess this variable must be set to `true` from the `goapp.yaml`, based on `tag-based-diff` and possibly other things?

bendiknesbo commented on internal/git/git.go:
```diff
@@ -80,6 +80,58 @@ func DiffToMain() ([]string, error) {
+func latestTag(pattern string) (string, error) {
+	// add exlucde on "*-*" removes all alpha/beta/rc etc from the list
+	return sh.Output("git", "describe", "--tags", "--abbrev=0", "--match", pattern, "--exclude", "*-*")
+}
```
In a repo without any existing tags, this may exit with the error `fatal: No names found, cannot describe anything.`

bendiknesbo replied:
The `--always` flag can be added to prevent it from failing, but then it's going to fall back to the commit hash of HEAD.

AtzeDeVries replied:
Yeah good point, i forget to add. I think that 'failing' on no tag is found is the actual correct behavior, else we might end up some silent fails.   

My initial idea to approach this would be that when we start using this a user/owner should set the initial tag. I could fall back to the initial commit in the main branch but then the changelog would be to big and cause issues down the line. 
I will look into the `--always` if that might fix the problem here.

AtzeDeVries replied:
The behavior is now

* Mage will look for Github releases starting with a certain prefix in the name of the release
* If no release found, it will use the diff based on the diff between main
* After a OCI is build a release is created

bendiknesbo reviewed (commented):
I like this approach, as it fixes a bug where several PR's being merged can lead to some changes not being built a docker-image for.

AtzeDeVries committed:
implement it based on github releases and fallback on main diff if no release found

Using github release allows us to use naming fo the release to do better
filtering.
Fallback to main if not release helps with onboarding.

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
make naming more uniform in gha

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix gha copilot comment

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries commented on internal/github/github.go:
```diff
@@ -158,3 +161,131 @@ func validateCommentBody(filename string) error {
+	url := fmt.Sprintf("%s/repos/%s/%s/releases", o.baseURL, o.owner, o.repo)
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	req.Header.Set("Authorization", "Bearer "+o.token)
```
fixed.

AtzeDeVries committed:
fix docs formatting change

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix docs

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
try this

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix techdocs validate

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
limit amount of commits in initial release

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
migrate some spaces to tabs

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

bendiknesbo commented on .github/workflows/reusable-goapp.yaml:
```diff
@@ -181,6 +186,45 @@ jobs:
+          GITHUB_TOKEN: ${{ steps.generate_token.outputs.token }}
+        with:
+          commitish: main
+          config-name: "release-drafter.yml"
```
What happens if this file is not present?

AtzeDeVries replied:
this will look at the release-drater in .github repository (shared). This is also how the generic workflow works https://github.com/coopnorge/github-workflow-release-drafter/blob/main/.github/workflows/release-drafter.yaml#L71

bendiknesbo commented on .github/workflows/reusable-goapp.yaml:
```diff
@@ -181,6 +186,45 @@ jobs:
+        id: generate_token
+        if: ${{ github.ref == 'refs/heads/main' && inputs.tag-based-diff }}
+        # v1.5.1
+        uses: actions/create-github-app-token@v2.2.1
```
Pin the action to a hash, and put the actual version in the comment after the hash

AtzeDeVries replied:
added.

bendiknesbo commented on .github/workflows/reusable-goapp.yaml:
```diff
@@ -181,6 +186,45 @@ jobs:
+        id: oci-tag
+        if: ${{ github.ref == 'refs/heads/main' && inputs.tag-based-diff }}
+        run: |
+          OCI_TAG=$(cat var/oci-images.json | yq -P | grep tag: | head -1 | yq .tag)
```
Can you quote `tag:` and `.tag`? Just to avoid it getting parsed as a shell built-in.

Also, can this return the string `null`?
We had some issues with this earlier, so we created some extra checks like these: https://github.com/coopnorge/idp-user-service/blob/main/.github/workflows/pr.yaml#L166-L170

AtzeDeVries replied:
i'll add the quote.

On line 205 i will check if the tag is empty and fail ( `-z $OCI_TAG`). We could maybe validate the tag to check if it matches the expected semver pattern.

bendiknesbo commented on internal/git/git.go:
```diff
@@ -81,6 +82,56 @@ func DiffToMain() ([]string, error) {
+	ref := "origin/main"
+	onMain, err := onMainBranch()
+	if err != nil {
+		return changedFiles, err
```
This confused me a bit, because I initially read it as: if on main-branch, return `changedFiles` (empty slice).

Can you wrap this error, and return nil for the slice?

```suggestion
		return nil, fmt.Errorf("failed to check if commit is on main branch: %w", err)
```

bendiknesbo commented on internal/git/git.go:
```diff
@@ -81,6 +82,56 @@ func DiffToMain() ([]string, error) {
+		// create release
+		if releaseRef != "" {
+			ref = releaseRef
+		}
```
Maybe add some validation if we are rerunning an older action? Semver tags are lexically ordered, so we could rely on that

AtzeDeVries replied:
i added that if the creation date of the release is newer/equal to than the current commit, then it will error. We are running this only on the main branch so commitdate should be safe i guess (no merge commit etc..)

bendiknesbo commented on internal/github/github.go:
```diff
@@ -158,3 +161,132 @@ func validateCommentBody(filename string) error {
+	}
+
+	for _, r := range releases {
+		if strings.HasPrefix(r.Name, prefix) {
+			return r.TagName, nil
+		}
```
Should we ignore draft and/or prereleases here?

AtzeDeVries replied:
mmm good point, i rewrote some stuff, now it will need to be caught here. I will also add tests for that.

bendiknesbo commented on internal/github/github.go:
```diff
@@ -158,3 +161,132 @@ func validateCommentBody(filename string) error {
+		return "", fmt.Errorf("failed to parse: %s\nerr: %w", string(body), err)
+	}
+
+	for _, r := range releases {
```
Is the releases guaranteed by the API to be in order?

AtzeDeVries replied:
actually no, but it usually is, but will order anyways.

bendiknesbo commented on internal/github/github_test.go:
```diff
@@ -0,0 +1,87 @@
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(tt.httpStatus)
+			_, err := w.Write([]byte(tt.payload))
+			_ = err
```
```suggestion
			assert.NoError(t, err)
```

AtzeDeVries replied:
added

bendiknesbo commented on internal/github/github_test.go:
```diff
@@ -0,0 +1,87 @@
+			_, err := w.Write([]byte(tt.payload))
+			_ = err
+		}))
+		defer server.Close()
```
This (these) defer(s) will not actually run until the entire test is done, so you will have `len(tests)` open HTTP servers open until every test is complete. 

This can quickly be fixed by running every unit-test inside a `t.Run` subtest, as then you create a new closure that the defer will run in.

AtzeDeVries replied:
added running in t.Run and doing server.Close within t.cleanup

bendiknesbo commented on internal/golang/golang.go:
```diff
@@ -112,8 +112,8 @@ func FindGoSourceCodeFolders(base string) ([]string, error) {
 // to the main branch
-func HasChanges(goSourceCodeFolders []string) (bool, error) {
-	changedFiles, err := git.DiffToMain()
+func HasChanges(goSourceCodeFolders []string, pattern string) (bool, error) {
+	changedFiles, err := git.DiffToTagPattern(pattern)
```
Question: Shouldn't this still run `DiffToMain` for users that haven't yet opted into the tag-based releases?

AtzeDeVries replied:
So opt out will always use the CHANGED_FILES env var override. 


https://github.com/coopnorge/mage/pull/414/changes#diff-4fccbfbac633897c14b167460e4d97b00076e3e9186d1ec2b51bc6ffb8b88088R84-R92

AtzeDeVries replied:
supplying the CHANGED_FILES as envvar will override any diff detection in code.

AtzeDeVries committed:
fix comments

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
add commit pin

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
ingest token

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
do diff on main based on dory path filters if no release is found

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix tag, docs and go lint

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
bump release-drater

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
fix date input

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
update docs a bit

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries committed:
add small doc

Signed-off-by: Atze de Vries <atze.wiebe.de.vries@coop.no>

AtzeDeVries commented:
example release 
https://github.com/coopnorge/helloworld/releases/tag/v2026.04.21105450
example renovate pr:
https://github.com/coopnorge/helloworld/pull/3365

bendiknesbo commented:
> example release [coopnorge/helloworld@`v2026.04.21105450` (release)](https://github.com/coopnorge/helloworld/releases/tag/v2026.04.21105450) example renovate pr: [coopnorge/helloworld#3365](https://github.com/coopnorge/helloworld/pull/3365)

Another benefit that I had missed, is that this will leverage GitHub's ability to detect when something is deployed (screenshot from https://github.com/coopnorge/helloworld/pull/3363):

<img width="935" height="222" alt="Screenshot 2026-04-21 at 15 55 09" src="https://github.com/user-attachments/assets/25dfaeac-ec2a-4369-95c3-3de8237dc0c2" />

bendiknesbo reviewed (approved):
This looks awesome to me :)

skasti commented on .github/workflows/goapp.yaml:
```diff
@@ -33,6 +33,13 @@ on:
         description: A comma separated list of Globs which triggers terraform ci
+      tag-based-diff:
+        type: boolean
+        default: true
```
You write that this is disabled by default, but make it enabled by default 😅
